File: components/auth/src/test/java/org/cloudfoundry/credhub/config/auth/OAuth2AuthenticationExceptionHandlerTest.java
package org.cloudfoundry.credhub.config.auth;

import org.springframework.mock.web.MockHttpServletRequest;
import org.springframework.mock.web.MockHttpServletResponse;
import org.springframework.security.authentication.InsufficientAuthenticationException;
import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
import org.springframework.security.oauth2.core.OAuth2Error;

import org.junit.Before;
import org.junit.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class OAuth2AuthenticationExceptionHandlerTest {

    private OAuth2AuthenticationExceptionHandler exceptionHandler;
    private MockHttpServletRequest request;
    private MockHttpServletResponse response;

    @Before
    public void setUp() {
        exceptionHandler = new OAuth2AuthenticationExceptionHandler();
        request = new MockHttpServletRequest();
        response = new MockHttpServletResponse();
    }

    @Test
    public void testOAuth2AuthenticationExceptionHandling() throws Exception {
        OAuth2Error error = new OAuth2Error(
                "error", "error-description", "error-uri");
        OAuth2AuthenticationException oae =
                new OAuth2AuthenticationException(error, "description");

        exceptionHandler.commence(request, response, oae);

        assertEquals(
                "{\"error\":\"error\",\"error_description\":\"error-description\"}",
                response.getContentAsString());
    }


    @Test
    public void testInsufficientAuthenticationExceptionHandling() throws Exception {
        InsufficientAuthenticationException iae
                = new InsufficientAuthenticationException("Full authentication required");

        exceptionHandler.commence(request, response, iae);

        assertEquals(
                "{\"error\":\"access_denied\",\"error_description\":\"Full authentication required\"}",
                response.getContentAsString());
    }
}

File: components/auth/src/test/kotlin/org/cloudfoundry/credhub/auth/OAuth2AuthenticationExceptionHandlerTest.kt
package org.cloudfoundry.credhub.auth

import com.jayway.jsonpath.JsonPath
import org.cloudfoundry.credhub.util.CurrentTimeProvider
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.mock.web.MockHttpServletRequest
import org.springframework.mock.web.MockHttpServletResponse
import org.springframework.security.authentication.InsufficientAuthenticationException
import org.springframework.security.oauth2.provider.token.store.jwk.JwkException
import java.security.cert.CertPathValidatorException
import javax.servlet.http.HttpServletRequest

internal class OAuth2AuthenticationExceptionHandlerTest {
    private lateinit var oAuth2AuthenticationExceptionHandler: OAuth2AuthenticationExceptionHandler
    private lateinit var request: HttpServletRequest
    private lateinit var response: MockHttpServletResponse

    @BeforeEach
    fun setUp() {
        oAuth2AuthenticationExceptionHandler = OAuth2AuthenticationExceptionHandler(CurrentTimeProvider())
        request = MockHttpServletRequest()
        response = MockHttpServletResponse()
    }

    @AfterEach
    fun tearDown() {
    }

    @Test
    fun handleException_whenGivenCertPathValidationException_addsErrorMessage() {
        val certPathValidatorException = CertPathValidatorException("this is our error")
        val jwkException = JwkException("some jwk error", certPathValidatorException)
        val exception = InsufficientAuthenticationException(certPathValidatorException.message, jwkException)

        oAuth2AuthenticationExceptionHandler.handleException(request, response, exception)

        assertThat(
            JsonPath.compile("error").read(response.contentAsString),
            equalTo("server_error")
        )
        assertThat(
            JsonPath.compile("error_description").read(response.contentAsString),
            equalTo("Server unable to communicate with backend UAA due to untrusted CA: this is our error")
        )
        assertThat(response.status, equalTo(500))
    }
}